internal/pkg/agent/application/gateway/fleet/fleet_gateway.go

```diff
@@ -155,38 +172,145 @@ func (f *fleetGateway) Run(ctx context.Context) error {
+
+					if checkinResult.err != nil {
+						var relaunchCheckinErr *needNewCheckinError
+						if errors.As(checkinResult.err, &relaunchCheckinErr) {
+							// checkin was cancelled because we have an updated state to send to fleet
+							state = relaunchCheckinErr.newState
+						}
+
+						continue
+					}
+
+					// if we got here the checkin has been completed
+					checkinResponse = checkinResult.response
+				}
+
+				actions := make([]fleetapi.Action, len(checkinResponse.Actions))
+				copy(actions, checkinResponse.Actions)
+				if len(actions) > 0 {
+					f.actionCh <- actions
+				}
+			}()
+
+		}
+	}
+}
+
+func (f *fleetGateway) performCancellableCheckin(ctx context.Context, initialState state.State, stateUpdates <-chan state.State, backoff backoff.Backoff, dbouncer debouncer) checkinResult {
+	checkinID, _ := uuid.NewV4()
+	checkinCtx, cancelCheckin := context.WithCancel(ctx)
+	checkinCtx = context.WithValue(checkinCtx, CheckinIDKey, checkinID.String())
+	f.log.Debugf("starting checkin %q", checkinID.String())
+	defer func() {
+		// make sure we don't leak contexts whatever happens
+		cancelCheckin()
+	}()
+
+	// Execute the checkin call asynchronously. For any errors returned by the fleet-server API
+	// the function will retry to communicate with fleet-server with an exponential delay and some
+	// jitter to help better distribute the load from a fleet of agents.
+	resCheckinChan := f.doExecuteAsync(checkinCtx, backoff, initialState)
+
+	// this variable will hold the latest state we got from coordinator
+	var updatedState *state.State
+	var debounceElapsed bool
 
-			// Execute the checkin call and for any errors returned by the fleet-server API
-			// the function will retry to communicate with fleet-server with an exponential delay and some
-			// jitter to help better distribute the load from a fleet of agents.
-			resp, err := f.doExecute(ctx, backoff)
-			if err != nil {
+	for {
+		select {
+		case <-ctx.Done():
+			return checkinResult{err: ctx.Err()}
+		case checkinResult := <-resCheckinChan:
+			return checkinResult
+		case <-dbouncer.Reached():
+			debounceElapsed = true
+			if updatedState != nil && !reflect.DeepEqual(updatedState, &initialState) {
+				// we detected that the updated State is different from the one of the current checkin, cancel and return the needNewCheckinError
+				cancelCheckin()
+				f.log.Debugf("cancelled checkin %s", checkinID.String())
+				cancelledCheckinRes := <-resCheckinChan
+				f.log.Debugf("reaped answer for cancelled checkin %q: %+v", checkinID, cancelledCheckinRes)
+				return checkinResult{err: &needNewCheckinError{newState: *updatedState}}
+			}
+		case newState := <-stateUpdates:
+			if !debounceElapsed {
+				f.log.Debugf(
+					"Updated state %+v received but current checkin %s not cancelled because we are still within debounce",
+					newState,
+					checkinID,
+				)
+				updatedState = &newState
+				_ = updatedState
```

Comment:
Why is this here?

Reply:
to avoid the linter screaming at me while I was developing it and the variable was unused. good catch!